src/index.ts

```diff
@@ -106,123 +75,88 @@ function initConfig(projectConfig: Forceable<Config>):
-        'were loaded before the trace agent was initialized:',
-        `[${modulesLoadedBeforeTrace.sort().join(', ')}]`);
-    // Stop storing these entries in memory
-    filesLoadedBeforeTrace.length = 0;
-    modulesLoadedBeforeTrace.length = 0;
+  if (!tracingSingleton) {
+    tracingSingleton = require('./tracing').tracing;
   }
 
   try {
-    // Initialize context propagation mechanism.
-    const m = config.clsMechanism;
-    const clsConfig: Forceable<TraceCLSConfig> = {
-      mechanism: m === 'auto' ? (useAH ? TraceCLSMechanism.ASYNC_HOOKS :
-                                         TraceCLSMechanism.ASYNC_LISTENER) :
-                                m as TraceCLSMechanism,
-      [FORCE_NEW]: config[FORCE_NEW]
-    };
-    cls.create(clsConfig, logger).enable();
-
-    traceWriter.create(config, logger).initialize((err) => {
-      if (err) {
-        stop();
-      }
-    });
-
-    traceAgent.enable(config, logger);
-
-    pluginLoader.create(config, logger).activate();
+    let tracing: Tracing;
+    try {
+      tracing = tracingSingleton.create(normalizedConfig, {});
+    } catch (e) {
+      // An error could be thrown if create() is called multiple times.
+      // It's not a helpful error message for the end user, so make it more
+      // useful here.
+      throw new Error('Cannot call start on an already created agent.');
+    }
+    tracing.enable();
+    tracing.logModulesLoadedBeforeTrace(filesLoadedBeforeTrace);
+    return tracingSingleton.get().traceAgent;
   } catch (e) {
-    logger.error(
-        'TraceAgent#start: Disabling the Trace Agent for the',
-        `following reason: ${e.message}`);
-    stop();
-    return traceAgent;
-  }
-
-  if (typeof config.projectId !== 'string' &&
-      typeof config.projectId !== 'undefined') {
-    logger.error(
-        'TraceAgent#start: config.projectId, if provided, must be a string.',
-        'Disabling trace agent.');
-    stop();
-    return traceAgent;
+    throw e;
+  } finally {
+    // Stop storing these entries in memory
+    filesLoadedBeforeTrace.length = 0;
   }
-
-  // Make trace agent available globally without requiring package
-  global._google_trace_agent = traceAgent;
-
-  logger.info('TraceAgent#start: Trace Agent activated.');
-  return traceAgent;
 }
 
+/**
+ * Get the previously created TraceAgent object.
+ * @returns An object exposing functions for creating custom spans.
+ */
 export function get(): PluginTypes.TraceAgent {
-  return traceAgent;
+  if (!tracingSingleton) {
+    tracingSingleton = require('./tracing').tracing;
+  }
+  if (tracingSingleton.exists()) {
+    return tracingSingleton.get().traceAgent;
+  } else {
+    // This code path maintains the current contract that calling get() before
+    // start() yields a disabled custom span API. It assumes that the use case
+    // for doing so (instead of returning null) is when get() is called in
+    // a file where it is unknown whether start() has been called.
```

Comment:
Hmmm.. this sounds odd. I don't understand the use case for this as you have described it. I don't think users would have any benefit of having a reference to a permanently disabled trace agent.

I think the reference we return from `get` will get 'enabled' once start is later called. I tested this theory out https://runkit.com/embed/3xg5kst085dt and it seems like my hunch is correct.

As such, this change is making an assumption that is incorrect.

Reply:
Ok, I figured out a way to circumvent this (I didn't like it anyway and was hoping someone had ideas about how to change it).